### PR TITLE
fix(22.8) Add processor to handle 'allocator' bug in Clickhouse

### PR DESCRIPTION
**Pull request by evanh**
There is some bug in Clickhouse 22.8 that can cause errors like "Too large size
(13837309855095848992) passed to allocator. It indicates an error."

The issue was narrowed down to conditions on low cardinality string columns,
and specific types of data in those columns. This processor hacks those
conditions to be in a form that allows the query to run.

### CODE REVIEW

File: snuba/query/processors/logical/condition_simplifier_processor.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    e.g. release IN tuple('a', 'b', 'c')
+    will cause an error.
+    THis processor does two things: if the rhs is a single value, change the condition
+    to be equals(). Otherwise, flip the sides using the `has` operator.
+    So has(tuple('a', 'b', 'c'), release) instead of in(release, tuple('a', 'b', 'c'))
+    """
+
+    def process_query(self, query: Query, query_settings: QuerySettings) -> None:
+        def transform_cond(exp: Expression) -> Expression:
+            print("CHECKING", exp)
```
Comment (onewland) [inline snuba/query/processors/logical/condition_simplifier_processor.py:25]: we want to remove this print, right?

Comment (evanh) [reply snuba/query/processors/logical/condition_simplifier_processor.py:25]: What! Why???

File: snuba/query/processors/logical/condition_simplifier_processor.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    22.8 has some problems when dealing with certain conditions on strings.
+    Specifically, a condition with multiple values on a string column,
+    e.g. release IN tuple('a', 'b', 'c')
+    will cause an error.
+    THis processor does two things: if the rhs is a single value, change the condition
+    to be equals(). Otherwise, flip the sides using the `has` operator.
+    So has(tuple('a', 'b', 'c'), release) instead of in(release, tuple('a', 'b', 'c'))
+    """
+
+    def process_query(self, query: Query, query_settings: QuerySettings) -> None:
```
Comment (volokluev) [inline snuba/query/processors/logical/condition_simplifier_processor.py:23]: do we want to put this behind a runtime config so we can turn on/off?

File: snuba/query/processors/logical/condition_simplifier_processor.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        def transform_cond(exp: Expression) -> Expression:
+            if not is_condition(exp):
+                return exp
+
+            assert isinstance(exp, FunctionCall)
+            if not exp.function_name == "in" or len(exp.parameters) != 2:
+                return exp
+
+            rhs = exp.parameters[1]
+            print(rhs)
```
Comment (volokluev) [inline snuba/query/processors/logical/condition_simplifier_processor.py:33]: another print :D

Comment (evanh) [reply snuba/query/processors/logical/condition_simplifier_processor.py:33]: It's a load bearing print.

**Review by volokluev — APPROVED** [review]
removing the print and feature flag would be good but I trust you with those

File: tests/test_discover_api.py [hunk 4]
```diff
@@ -747,7 +747,7 @@ def test_transaction_group_ids(self) -> None:
                 entity="discover_transactions",
             ).data
         )
-
+        print("RESULT", result)
```
Comment (volokluev) [inline tests/test_discover_api.py:750]: 👀

File: tests/test_snql_api.py [hunk 5]
```diff
@@ -1391,6 +1391,75 @@ def test_hexint_in_condition(self) -> None:
...
+        )
+
+    def test_discover_semver_bug(self) -> None:
+        response = self.post(
+            "/discover/snql",
+            data=json.dumps(
+                {
+                    "consistent": False,
+                    "turbo": False,
+                    "query": "MATCH (discover) SELECT event_id AS `id` WHERE release IN array('test@1.2.3+123', 'test2@1.2.4+124') AND timestamp >= toDateTime('2024-04-18T22:28:40.928000') AND timestamp < toDateTime('2024-04-19T22:28:40.306000') AND project_id IN array(4553863597588481) LIMIT 50",
```
Comment (volokluev) [inline tests/test_snql_api.py:1445]: @evanh this is the sentry test that's failing in a bizarre fashion:

```
'DB::Exception: First argument for function has must be an array.. 
```

sql:

```
SELECT (replaceAll(toString(event_id), '-', '') AS _snuba_event_id) FROM discover_local 
PREWHERE in((project_id AS _snuba_project_id), [4553863597588481]) 
WHERE equals(deleted, 0) AND has(['test@1.2.3+123', 'test2@1.2.4+124'], (release AS _snuba_release)) 
AND greaterOrEquals((timestamp AS _snuba_timestamp), toDateTime('2024-04-18T22:28:40', 'Universal')) 
AND less(_snuba_timestamp, toDateTime('2024-04-19T22:28:40', 'Universal')) LIMIT 50 OFFSET 0
```

Comment (evanh) [reply tests/test_snql_api.py:1445]: Can you replace all the hardcoded values here with test values (project_id, org_id, timestamp etc.)?

Comment (volokluev) [reply tests/test_snql_api.py:1445]: these are the values that are in the failing sentry test

Comment (volokluev) [reply tests/test_snql_api.py:1445]: Just tried it, it fails the same way

Comment (evanh) [reply tests/test_snql_api.py]: This is very weird, because the SQL query works in production and when running it directly on Clickhouse.

Comment (evanh) [reply tests/test_snql_api.py]: Had a (hacky) idea: I just removed the optimizer from the discover entity. We don't actually need it there, just discover_transactions and transactions.

**Review by volokluev — APPROVED** [review]

**Pull request merged**